### PR TITLE
remove _hidden_indices() from LocationEncoder

### PR DESCRIPTION
**Pull request by jeff-regier**
This is a step towards resolving #540. It builds on PR #545, removing the complex `_get_hidden_indices()` function from `LocationEncoder`. The main innovation is to cast `per_source_params` as a 5-dimensional tensor that can be sliced by detection. That way all the parameters for a particular detection can be referenced without  storing their indices in a tensor. Additionally, all the parameters for a particularly parameter group (e.g., `loc_mean`) can also be referenced without an index lookup table, by slicing a different dimension of the 5d `per_source_params` tensor.

### CODE REVIEW

File: tests/test_encoder_forward.py [hunk 1]
```diff
@@ -50,13 +46,18 @@ def test_forward(self, devices):
...
+
             images *= background_tensor.sqrt()
             images += background_tensor
             var_params = star_encoder.encode(images, background_tensor)
-            star_encoder.max_a_post(var_params)
+            catalog = star_encoder.max_a_post(var_params)
+
+            assert catalog.n_sources.size() == torch.Size([batch_size, n_tiles_h, n_tiles_w])
+            correct_locs_shape = torch.Size([batch_size, n_tiles_h, n_tiles_w, max_detections, 2])
+            assert catalog.locs.shape == correct_locs_shape
```
Comment (jeff-regier) [inline tests/test_encoder_forward.py:60]: These new assert statements pass on master too

Comment (dereklhansen) [reply tests/test_encoder_forward.py:60]: Looks good!

File: bliss/models/location_encoder.py [hunk 2]
```diff
@@ -352,28 +358,32 @@ def _encode_for_n_sources(
...
+        ts5 = ts4.expand(ts4.size(0), ts4.size(1), self.n_params_per_source)
+
+        # next, we pad `params_per_source` with a dummy column of zeros that will be looked up
+        # (copied) whenever fewer the `max_detections` sources are present. `gather` does the copy.
+        pps2 = rearrange(params_per_source, "b nth ntw td pps -> (b nth ntw) td pps")
+        pps3 = F.pad(pps2, (0, 0, 0, 1))
+        pps4 = torch.gather(pps3, 1, ts5)
+        pps5 = rearrange(pps4, "np (ns md) pps -> ns np md pps", ns=tile_n_sources.size(0))
+
+        # finally, we slice pps5 by parameter group because these groups are treated differently,
```
Comment (dereklhansen) [inline bliss/models/location_encoder.py:377]: Could we just use `torch.split()` for this part? E.g.
```
dims = param.values()
var_params_split = torch.split(pp5, dims, -1)
var_params_for_n_sources = {k: v for k,v in zip(dims.keys(), var_params_split)}
```

**Review by dereklhansen — APPROVED** [review]
LGTM. I just have a few comments on some naming

**Comment by dereklhansen** [comment]
Looks like the remaining errors are due to mismatch in LocationEncoder state-dict on the VAE

**Comment by dereklhansen** [comment]
- [ ] Retrain VAE weights

**Comment by jeff-regier** [comment]
I retrained the location encoder in `case_studies/sdss_galaxies`. 

For the retraining the location encoder in `case_studies/sdss_galaxies_vae` fails due to an out-of-memory error. This happens for me on master too, so I don't think it's an issue with the PR.

File: bliss/models/location_encoder.py [hunk 3]
```diff
@@ -352,28 +358,32 @@ def _encode_for_n_sources(
             `n_samples x n_ptiles x max_detections x ...`
         """
         tile_n_sources = tile_n_sources.clamp(max=self.max_detections)
-        n_ptiles = var_params_flat.size(0)
-        vpf_padded = F.pad(var_params_flat, (0, self.dim_out_all, 0, 0))
 
+        # first, we transform `tile_n_sources` so that it can be used as an index
+        # for looking up detections in `params_per_source`
+        ts2 = rearrange(tile_n_sources, "ns b nth ntw -> ns (b nth ntw)")
```
[the remark this answers is not in the document]
Comment (jeff-regier) [reply bliss/models/location_encoder.py:365]: I think `ns` is used elsewhere in this file to mean `n_samples`. Is it being used to mean `n_sources` too?

File: bliss/models/location_encoder.py [hunk 4]
```diff
@@ -126,10 +126,14 @@ def __init__(
...
-        # get indices into the triangular array of returned parameters
-        indx_mats = self._get_hidden_indices()
-        for k, v in indx_mats.items():
-            self.register_buffer(k + "_indx", v, persistent=False)
+        # the next block of code constructs `self.n_detections_map`, which is a 2d tensor with
+        # size (self.max_detections + 1, self.max_detections).
+        # Each row corresponds to a number of detections in a tile (including zero).
+        # Each row contains the indices of the relevant detections, padded by a dummy value.
+        md, ntd = self.max_detections, self.n_total_detections
+        self.n_detections_map = torch.full((md + 1, md), ntd, device=self.device)
```
[the remark this answers is not in the document]
Comment (jeff-regier) [reply bliss/models/location_encoder.py]: Yep, that's it

File: bliss/models/location_encoder.py [hunk 5]
```diff
@@ -126,10 +126,14 @@ def __init__(
...
         self.log_softmax = nn.LogSoftmax(dim=1)
 
-        # get indices into the triangular array of returned parameters
-        indx_mats = self._get_hidden_indices()
-        for k, v in indx_mats.items():
-            self.register_buffer(k + "_indx", v, persistent=False)
+        # the next block of code constructs `self.n_detections_map`, which is a 2d tensor with
+        # size (self.max_detections + 1, self.max_detections).
+        # Each row corresponds to a number of detections in a tile (including zero).
+        # Each row contains the indices of the relevant detections, padded by a dummy value.
```
[the remark this answers is not in the document]
Comment (jeff-regier) [reply bliss/models/location_encoder.py:132]: Not really: the row indexed by two, for example, corresponds to the case where a tile has two sources. The values in the row would be `[1, 2, 10, 10]` (assuming `max_detections = 4`) . It's pretty tricky to describe this structure clearly in comments, unfortunately

File: bliss/models/location_encoder.py (hunk 2 above)
Comment (jeff-regier) [reply bliss/models/location_encoder.py:377]: I like it!

File: bliss/models/location_encoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/models/location_encoder.py:365]: > When I try to re-train the location encoder in `case_studies/sdss_galaxies_vae`, I get an out-of-memory error. This happens for me on master too, so I don't think it's an issue with the PR.

I don't have this issue on my end

**Comment by dereklhansen** [comment]
> When I try to re-train the location encoder in `case_studies/sdss_galaxies_vae`, I get an out-of-memory error. This happens for me on master too, so I don't think it's an issue with the PR.

There's three environment variables in my Makefile to set the GPU (to allow for parallel training). Specifically, you need to set GPU_SLEEP to the GPU you want to use for the location encoder.
```
make GPU_SLEEP=1 models/sdss_location_encoder.pt
```

File: bliss/models/location_encoder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/models/location_encoder.py:132]: Gotcha, I understand it now. I can't really think of a better way to explain it either.

**Comment by jeff-regier** [comment]
I tested this PR with `pytest --gpu` and it triggers no new test failures. The test failures are the same as the failures on master, which I reported in issue #563.

**Pull request merged**